File: code/Java/state/gumballmachine/SoldOutState.java
package state.gumballmachine;

public class SoldOutState implements State{
    private GumballMachine gumballMachine;

    public SoldOutState(GumballMachine gm) {
        this.gumballMachine = gm;
    }

    @Override
    public void insertQuarter() {
        System.out.println("You can't insert a quarter, the machine is sold out");
    }

    @Override
    public void ejectQuarter() {
        System.out.println("You can't eject, you have not inserted a quarter yet");
    }

    @Override
    public void turnCrank() {
        System.out.println("You turned, but there is no quarter");
    }

    @Override
    public void dispense() {
        System.out.println("No gumball dispense");
    }

    @Override
    public void refill() {
        if(gumballMachine.getCount() > 0)
            gumballMachine.setCurState(gumballMachine.getNoQuarterState());
        else
            gumballMachine.setCurState(gumballMachine.getSoldOutState());
    }

//    @Override
//    public String toString() {
//        return "SoldOutState{" +
//                "gumballMachine=" + gumballMachine +
//                '}';
//    }
}

File: code/C++/state/soldoutstate.cpp
#include "soldoutstate.h"

SoldOutState::SoldOutState(GumballMachine * const gm) :gumballMachine(gm) {}

void SoldOutState::insertQuarter()
{
	std::cout << "You can't insert a quarter, the machine is sold out" << std::endl;
}

void SoldOutState::ejectQuarter()
{
	std::cout << "You can't eject, you have not inserted a quarter yet" << std::endl;
}

void SoldOutState::turnCrank()
{
	std::cout << "You turned, but there is no quarter" << std::endl;
}

void SoldOutState::dispense()
{
	std::cout << "No gumball dispense" << std::endl;
}

void SoldOutState::refill()
{
	if (this->gumballMachine->getCount() > 0)
		this->gumballMachine->setState(this->gumballMachine->getNoQuarterState());
	else
		this->gumballMachine->setState(this->gumballMachine->getSoldOutState());
}

void SoldOutState::display()
{
	std::cout << "SoldOutState, count:" << this->gumballMachine->getCount() << std::endl;
}